src/pint/fitter.py

```diff
@@ -1287,10 +1265,140 @@ def fit_toas(
+                maxiter=maxiter,
+                required_chi2_decrease=required_chi2_decrease,
+                max_chi2_increase=required_chi2_decrease,
+                min_lambda=required_chi2_decrease,
+                debug=debug,
+            )
+        else:
+            log.debug("Will fit for noise parameters.")
+            for _ in range(noise_fit_niter):
+                self._fit_toas(
+                    maxiter=maxiter,
+                    required_chi2_decrease=required_chi2_decrease,
+                    max_chi2_increase=max_chi2_increase,
+                    min_lambda=min_lambda,
+                    debug=debug,
+                )
+                values, errors = self._fit_noise()
+                self._update_noise_params(values, errors)
+
+            return self._fit_toas(
+                maxiter=maxiter,
+                required_chi2_decrease=required_chi2_decrease,
+                max_chi2_increase=max_chi2_increase,
+                min_lambda=min_lambda,
+                debug=debug,
+            )
+
     @property
     def fac(self):
         return self.current_state.fac
 
+    def _get_free_noise_params(self):
+        """Returns a list of all free noise parameters."""
+        return [
+            fp
+            for fp in self.model.get_params_of_component_type("NoiseComponent")
+            if not getattr(self.model, fp).frozen
+        ]
+
+    def _update_noise_params(self, values, errors):
+        """Update the model using estimated noise parameters."""
+        free_noise_params = self._get_free_noise_params()
+        for fp, val, err in zip(free_noise_params, values, errors):
+            getattr(self.model, fp).value = val
+            getattr(self.model, fp).uncertainty_value = err
+
+    def _fit_noise(self):
+        """Estimate noise parameters and their uncertainties. Noise parameters
+        are estimated by numerically maximizing the log-likelihood function including
+        the normalization term. The uncertainties thereof are computed using the
+        numerically-evaluated Hessian."""
+        free_noise_params = self._get_free_noise_params()
+
+        xs0 = [getattr(self.model, fp).value for fp in free_noise_params]
+
+        model1 = copy.deepcopy(self.model)
+
+        def _mloglike(xs):
+            """Negative of the log-likelihood function."""
+            for fp, x in zip(free_noise_params, xs):
+                getattr(model1, fp).value = x
+
+            res = Residuals(self.toas, model1)
+            chi2, lognorm = res.calc_chi2(lognorm=True)
+
+            return chi2 / 2 + lognorm
+
+        result = opt.minimize(_mloglike, xs0, method="Nelder-Mead")
+
+        def _like(xs):
+            """The likelihood function normalized by its maximum value.
+            Note that this function will give sensible results only in the
+            vicinity of the maximum-likelihood point and will overflow
+            otherwise due to the `exp()` function."""
```

Comment:
Added exception.

Reply:
Since this is a nested function, it is not available to the user. Within the outer function, it is called by `Hessdiag` to evaluate the Hessian numerically *after the fitting has been done*. Hessdiag by definition will call `_like` in the vicinity of the maximum likelihood value. 

The fitting itself is done by minimizing the `_mloglike` function, which only computes the log-likelihood and not the likelihood.

Reply:
OK... It turns out that I don't need _like at all. Rather, the uncertainties can be computed by taking the Hessian of _mloglike itself. So I have removed _like altogether and there is no overflow issue now.